src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+  }
+  const common::TransactionConfig &config = config_or_status.value();
+  *response->mutable_config() = config;
+  const SplitTransaction split_transaction =
+      SplitClientTransaction(request->transaction());
+  if (split_transaction.transactions.empty()) {
+    return grpc::Status(grpc::INVALID_ARGUMENT, "No operations found");
+  }
+  if (RequiresBlockchain(split_transaction)) {
+    absl::Status block_chain_status =
+        StartVoting(transaction_id, config.presumed_abort_time(),
+                    split_transaction.transactions.size());
+    if (!block_chain_status.ok()) {
+      return utils::FromAbslStatus(block_chain_status,
+                                   "Failed to start voting in blockchain");
+    }
+  }
+  cohort::PrepareTransactionRequest prepare_request;
+  prepare_request.set_transaction_id(transaction_id);
+  *prepare_request.mutable_config() = config;
+  if (split_transaction.transactions.size() == 1) {
+    prepare_request.set_only_cohort(true);
+    metadata.single_cohort_namespace = split_transaction.transactions[0].ns;
+  }
+  size_t cohort_index = 0;
+  for (const TransactionWithNamespace &transaction_with_namespace :
+       split_transaction.transactions) {
+    if (split_transaction.transactions.size() != 1) {
+      prepare_request.set_cohort_index(cohort_index);
+    }
+    *prepare_request.mutable_transaction() =
+        transaction_with_namespace.transaction;
+    if (transaction_with_namespace.transaction_type != WRITE_ONLY_TRANSACTION) {
+      metadata.get_cohort_namespaces.push_back(transaction_with_namespace.ns);
+    }
+    // If this is the last cohort, then it may have been sent to all cohorts.
+    if (cohort_index == split_transaction.transactions.size() - 1) {
+      metadata.possibly_sent_to_all_cohorts = true;
+    }
+    // TODO(benjmarks22): Retry failed RPCs.
+    // For now the transactions with failed RPCs will abort at the blockchain
+    // at the presumed abort time.
+    PrepareCohortTransaction(transaction_id, transaction_with_namespace.ns,
+                             prepare_request, *context);
+    ++cohort_index;
+  }
+  return grpc::Status::OK;
+}
+
+absl::Status CoordinatorServer::StartVoting(
+    const std::string &transaction_id,
+    const google::protobuf::Timestamp &presumed_abort_time,
+    size_t num_cohorts) {
+  const time_t abort_time =
+      absl::ToTimeT(absl::FromUnixSeconds(presumed_abort_time.seconds()) +
+                    absl::Nanoseconds(presumed_abort_time.nanos()));
+  return Blockchain::StartVoting(transaction_id, abort_time, num_cohorts);
+}
+
+absl::StatusOr<Blockchain::VotingDecision> CoordinatorServer::GetVotingDecision(
+    const std::string &transaction_id) {
+  return Blockchain::GetVotingDecision(transaction_id);
+}
+
+absl::Time CoordinatorServer::Now() { return absl::Now(); }
+
+void CoordinatorServer::PrepareCohortTransaction(
+    const std::string &transaction_id, const Namespace &ns,
```

Comment:
'ns' is too short which triggers readability-identifier-length warning on my editor - maybe we don't use addr here.

Reply:
Changed to namespace_.